[CIR] Upstream basic support for ArrayType

AmrDeveloper opened the pull request:
This change adds the basic support for ArrayType

Issue #130197

AmrDeveloper committed:
[CIR] Upstream basic support for ArrayType

AmrDeveloper commented:
There will be two upcoming patches.

1 - Initializing const array and zero initializing support (Depends on upstreaming `ConstantEmitter` and `ZeroAttr`)
2 - Dynamic-size arrays, and accessing arrays using the `cir.ptr_stride` operation (Depends on upstreaming `CastOp`)

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenBuilder.h:
```diff
@@ -33,6 +33,14 @@ class CIRGenBuilderTy : public cir::CIRBaseBuilderTy {
+    if (mlir::isa<cir::PointerType, cir::ArrayType, cir::BoolType,
+                  cir::IntType>(ty))
+      return true;
+    assert(0 && "Unimplemented size for type");
```
looks like we've been doing `llvm_unreachable` for not yet implemented.  That said, its pretty weird that asking this question (`isSized`) is a not-implemented?  What sort of stuff is missing here that we cannot just answer this question?

andykaylor replied:
I think it's that the types are not implemented, so we can't return an answer for a type that doesn't exist yet.

AmrDeveloper replied:
I am thinking to make `isSized` simple 

`return mlir::isa<cir::PointerType, cir::ArrayType, cir::BoolType, cir::IntType>(ty)`

And when I call it I can assert or errorNYI

andykaylor replied:
I think it's better to have the "NYI" indication in the `isSize` definition since that's where the code will need to be updated to add support for other types.

AmrDeveloper replied:
I think if we put assert on isSize we can't recover from calling isSize, for example currently I recovering by putting sint32 and 0 size 🤔, what do you think?

erichkeane replied:
I like the idea of having `isSized` do an NYI error, then returning a 'sane' value (like false) instead of the assert.

That way, we don't really have to do any real checking in its use, just call 'isSized' or `assert(isSized())` in the use.

erichkeane replied:
Hmm... that is unfortunate.  It wouldn't be novel to have it take CGM as an argument, but I don't know what appetite there is for that.

AmrDeveloper replied:
`It wouldn't be novel to have it take CGM as an argument` exacily, I think in future we should move `errorNTI` to be a helper function that takes DiagnosticEngine as argument and from any CIEGen we can call it, even in `CIEGenBuilder` we can get Diag from context

erichkeane replied:
Still is an assert

andykaylor replied:
I think an assert is appropriate here. Perhaps the wording could be changed. This should be returning an answer for every type that we expect to get here. We don't expect non-CIR mlir types, like mlir::Integer.
```suggestion
    assert(0 && "Unexpected MLIR type");
```

erichkeane replied:
I think it isn't quite that.... there is an NYI state here as well.  If we dont' expect non-CIR MLIR types, an unreachable would be appropriate here instead.  IF all CIR-MLIR types are sized, asking this question seems silly.

andykaylor replied:
Hmm. I just looked again at the incubator implementation. I thought vector types returned false here, but they return `isSized(elementType)`, so nothing implemented there returns false without going through the assert. I compared the incubator code to the classic codegen, and the classic codegen is calling llvm::Type::isSized. Looking into that, it seems to be things like functions, labels, and target extension types that return false. So, you're right that there is something not yet implemented here, but the thing that isn't implemented is the existence of types that need special handling, so we can't even check for this being the thing that isn't implemented.

Maybe the thing to do here is get rid of `CIRGenBuilder::isSized()` completely and add `assert(!cir::MissingFeature::NonSizedTypes());` at the call site. This seems to only be used in two places in the incubator and both will have the same issue.

erichkeane replied:
I could live with that.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenTypes.cpp:
```diff
@@ -202,6 +202,18 @@ mlir::Type CIRGenTypes::convertType(QualType type) {
+
+    // FIXME: In LLVM, "lower arrays of undefined struct type to arrays of
+    // i8 just to have a concrete type". Not sure this makes sense in CIR yet.
+    assert(builder.isSized(elemTy) && "not implemented");
```
This should probably use the `cgm.errorNYI`, then just come up with a way to recover.  That said, this is exactly why having `isSized` assert is awkward here.

AmrDeveloper replied:
Okey, thats make sense and i will try to recover with 0 size

andykaylor commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -166,6 +166,9 @@ static LogicalResult checkConstantTypes(mlir::Operation *op, mlir::Type opType,
     return success();
   }
 
+  if (mlir::isa<cir::ArrayType>(opType))
```
I don't see this in the incubator implementation of this function. Why is this here?

andykaylor commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -369,6 +369,22 @@ BoolType::getABIAlignment(const ::mlir::DataLayout &dataLayout,
+//===----------------------------------------------------------------------===//
+
+llvm::TypeSize
+ArrayType::getTypeSizeInBits(const ::mlir::DataLayout &dataLayout,
```
Are these functions being called with any of the test cases? I don't see any checks that would be related to this.

AmrDeveloper replied:
I was thinking of `sizeof` operator but it's not implemented yet, is there is another way to test the size in lit test

andykaylor replied:
My point was that if the code isn't being called, you should leave it out until it is needed.

AmrDeveloper replied:
Yes, but it's part of the `DataLayoutTypeInterface` interface so we must implement it, I can test it after merging the sizeof PR https://github.com/llvm/llvm-project/pull/130847

andykaylor replied:
I see. Thanks for the clarification.

andykaylor reviewed (commented):
Can you also add a test for lowering to LLVM IR?

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -280,6 +280,25 @@ def CIR_BoolType :
+  let parameters = (ins "mlir::Type":$eltType, "uint64_t":$size);
+
+  let assemblyFormat = [{
+    `<` $eltType `x` $size `>`
```
Please add CIR to CIR tests to exercise the parser

AmrDeveloper replied:
Done

AmrDeveloper committed:
Add test for LLVM lowering, CIR test and address comments

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenTypes.cpp:
```diff
@@ -202,6 +202,21 @@ mlir::Type CIRGenTypes::convertType(QualType type) {
+    mlir::Type elemTy = convertTypeForMem(arrTy->getElementType());
+
+    if (!builder.isSized(elemTy)) {
+      cgm.errorNYI(SourceLocation(), "unimplemented size for type", type);
```
I don't like the error being handled this way. If you're going to report an error about unimplemented types here, the check for the types should be here. If someone were to change the implementation of isSized(), they would have no way to know that this code needed to be updated.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuilder.h:
```diff
@@ -33,6 +33,11 @@ class CIRGenBuilderTy : public cir::CIRBaseBuilderTy {
   }
+
+  bool isSized(mlir::Type ty) {
+    return mlir::isa<cir::PointerType, cir::ArrayType, cir::BoolType,
```
I think this was better with the assert for other types. This function should handle all types that can occur in the current form of CIR. As we add new types, this function should be updated to indicate if they are sized or not. So, this isn't quite like other `errorNYI` cases. An assert here is reasonable as a permanent part of the code.

AmrDeveloper replied:
Yes i will update it, and i think no need for NYI or other assert when calling it because in call cases we will not need to recover

andykaylor commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -572,6 +572,10 @@ static void prepareTypeConverter(mlir::LLVMTypeConverter &converter,
     return mlir::LLVM::LLVMPointerType::get(type.getContext(), targetAS);
   });
+  converter.addConversion([&](cir::ArrayType type) -> mlir::Type {
+    auto ty = convertTypeForMemory(converter, dataLayout, type.getEltType());
```
```suggestion
    mlir::Type ty = convertTypeForMemory(converter, dataLayout, type.getEltType());
```

andykaylor commented on clang/test/CIR/CodeGen/array.cpp:
```diff
@@ -0,0 +1,24 @@
+// CHECK: cir.global external @a : !cir.array<!cir.int<s, 32> x 10>
+
+int aa[10][10];
+// CHECK: cir.global external @aa : !cir.array<!cir.array<!cir.int<s, 32> x 10> x 10>
```
Use different sizes for the dimensions here so we can verify that the CIR generated has them in the right order.

andykaylor commented on clang/test/CIR/CodeGen/array.cpp:
```diff
@@ -0,0 +1,24 @@
+}
+
+void f2(int p[10]) {}
+// CHECK: cir.func @f2(%arg0: !cir.ptr<!cir.int<s, 32>>
```
Can you also check the `cir.alloca` generated in this case?

AmrDeveloper committed:
Address code review comments

andykaylor commented on clang/test/CIR/Lowering/array.cpp:
```diff
@@ -0,0 +1,24 @@
+void f() {
+  int l[10];
+}
+// CHECK: alloca [10 x i32], i64 1, align 16
```
I think it's a good practice to have a check for the function name to establish that we're seeing the `alloca` in the correct context.

andykaylor reviewed (commented):
I have just one minor comment on the update. Please wait for @erichkeane to review again, but this looks good to me.

AmrDeveloper committed:
Update the test to check function name of alloca

erichkeane reviewed (commented):
My concern is still not fixed, and Andy/Bruno still have comments here, so I don't believe this to be ready yet.

AmrDeveloper commented:
> My concern is still not fixed, and Andy/Bruno still have comments here, so I don't believe this to be ready yet.

I already addressed all comments except the test for Size and it's need sizeof PR to merge, and the assert line which we need to agree on, other than that every thing is addressed :D

AmrDeveloper committed:
Update assert message

Co-authored-by: Andy Kaylor <akaylor@nvidia.com>

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenBuilder.h:
```diff
@@ -33,6 +33,15 @@ class CIRGenBuilderTy : public cir::CIRBaseBuilderTy {
+                  cir::IntType>(ty))
+      return true;
+
+    assert(0 && "Unexpected MLIR type");
```
I think the 
```suggestion
    assert(!MissingFeatures::unsizedTypes());
```

was the suggestion here (see the other uses of MissingFeatures).

AmrDeveloper replied:
Okey, i will check MissingFeatures and update it

erichkeane reviewed (approved):
1 nit, else LGTM.

AmrDeveloper committed:
Add MissingFeatures for unsized types

AmrDeveloper committed:
Merge branch 'main' into cir_upstream_array_type

AmrDeveloper committed:
Fix code format after merging
